### ISSUE TITLE
Harbor 2.0 without HTTPS - Issue to authenticate (htpasswd)  behind a reverse proxy 

### ISSUE THREAD

**Issue by PauloFavero** [goharbor/harbor#12364]
Hi everyone,

While installing Harbor 1.10, I faced an authentication issue to login into docker registry as follows: 

`Error response from daemon: Get https://my.registry.company.com/v2/: unauthorized: authentication required`

I found that this was due to be running Harbor without HTTPS behind a reverse proxy. After some research, I successfully got the version 1.10 working behind a reverse proxy doing this [workaround](https://github.com/goharbor/harbor/issues/3114#issuecomment-432962664). It means that a changed `relm: http to relm: https`

Everything was working nicely until I updated to version 2.0 and started to get the same authentication error. 
At first glance, I went to harbor/common/config/registry/config.yml to do the same workaround as before but, the authentication method has changed as it shows in this [commit](https://github.com/goharbor/harbor/commit/2064a1cd6d9b97fd2c7445885cdc4eb461455af9) and I am not able to fix it anymore. 

**Authentication method (Harbor 1.10.0)**
```
auth:
  token:
    issuer: harbor-token-issuer
    realm: https://my.registry.company.com/service/token
    rootcertbundle: /etc/registry/root.crt
    service: harbor-registry
```
**Authentication method (Harbor 2.0.0)**
```
  htpasswd:
     realm: harbor-registry-basic-realm
     path: /etc/registry/passwd
```

Is there a way in which I can use the token authentication method in version 2.0.0? Or maybe another way to make harbor without HTTPS authenticate behind a reverse proxy?

Thanks

**Comment by reasonerjt** [comment]
@PauloFavero could you curl the `https://xxxx/v2` url and let us know the response headers?

**Comment by PauloFavero** [comment]
Hi @reasonerjt,

Thanks for your swift reply.

I did the same curl test as [here](https://github.com/goharbor/harbor/issues/3114#issuecomment-424992795)

First, I tried to simulate the login with:

`curl -u admin:password "https://my.registry.company.com/service/token?account=admin&client_id=docker&offline_token=true&service=harbor-registry"`

And I got the following output:

```
{
  "token": "xxxx",
  "access_token": "",
  "expires_in": 1800,
  "issued_at": "2020-07-06T15:31:36Z"
}
```

Then, I did the requested curl command : 

`curl -X GET -I "https://my.registry.company.com/v2/"`

Output:

```
HTTP/1.1 401 Unauthorized
Server: nginx/1.14.0 (Ubuntu)
Date: Mon, 06 Jul 2020 15:32:08 GMT
Content-Type: application/json; charset=utf-8
Content-Length: 76
Connection: keep-alive
Docker-Distribution-Api-Version: registry/2.0
Set-Cookie: sid=228ff16b93d69d65b788ff9495f47c5e; Path=/; HttpOnly
Www-Authenticate: Bearer realm="http://my.company.com/service/token",service="harbor-registry"
X-Request-Id: 6bc7af8c-18c5-42c8-aa6a-5f5e21d7fef8
Strict-Transport-Security: max-age=31536000; includeSubDomains
```

Instead of HTTPS, I tried to curl with HTTP:

`curl -X GET -I "http://registry.kascosys.com.br/v2/"`

Output:

```
HTTP/1.1 301 Moved Permanently
Server: nginx/1.14.0 (Ubuntu)
Date: Mon, 06 Jul 2020 15:32:40 GMT
Content-Type: text/html
Content-Length: 194
Connection: keep-alive
Location: https://my.registry.company.com/v2/
```

Docker Compose Version: 1.26.1, build f216ddbf
Docker Version: 19.03.12, build 48a66213fe

**Comment by PauloFavero** [comment]
Hi @reasonerjt,

Thanks for your swift reply.

I did the same curl test as [here](https://github.com/goharbor/harbor/issues/3114#issuecomment-424992795)

First, I tried to simulate the login with:

`curl -u admin:password "https://my.registry.company.com/service/token?account=admin&client_id=docker&offline_token=true&service=harbor-registry"`

And I got the following output:

```
{
  "token": "xxxx",
  "access_token": "",
  "expires_in": 1800,
  "issued_at": "2020-07-06T15:31:36Z"
}
```

Then, I did the requested curl command : 

`curl -X GET -I "https://my.registry.company.com/v2/"`

Output:

```
HTTP/1.1 401 Unauthorized
Server: nginx/1.14.0 (Ubuntu)
Date: Mon, 06 Jul 2020 15:32:08 GMT
Content-Type: application/json; charset=utf-8
Content-Length: 76
Connection: keep-alive
Docker-Distribution-Api-Version: registry/2.0
Set-Cookie: sid=228ff16b93d69d65b788ff9495f47c5e; Path=/; HttpOnly
Www-Authenticate: Bearer realm="http://my.company.com/service/token",service="harbor-registry"
X-Request-Id: 6bc7af8c-18c5-42c8-aa6a-5f5e21d7fef8
Strict-Transport-Security: max-age=31536000; includeSubDomains
```

Instead of HTTPS, I tried to curl with HTTP:

`curl -X GET -I "http://my.registry.company.com/v2/"`

Output:

```
HTTP/1.1 301 Moved Permanently
Server: nginx/1.14.0 (Ubuntu)
Date: Mon, 06 Jul 2020 15:32:40 GMT
Content-Type: text/html
Content-Length: 194
Connection: keep-alive
Location: https://my.registry.company.com/v2/
```

Docker Compose Version: 1.26.1, build f216ddbf
Docker Version: 19.03.12, build 48a66213fe

**Comment by reasonerjt** [comment]
@PauloFavero 

In v2 the value of realm in the header `Www-Authenticate` is determined by this line:
https://github.com/goharbor/harbor/blob/master/src/server/middleware/v2auth/auth.go#L121

**Comment by PauloFavero** [comment]
@reasonerjt 

I didn't get what to do. Do I have to change something in this code and then build the container?

**Comment by rstuberg-ib** [comment]
I am having  the same issue (helm v2.0.0) however my header is being set as
```www-authenticate: Basic realm="harbor"```

When trying the same query as
```curl -X GET -I "http://my.registry.company.com/v2/"```

What configuration file or log can we look at to debug what is being set?

**Comment by PauloFavero** [comment]
Any updates on this issue?

**Comment by tomhouweling1987** [comment]
@PauloFavero is followed the steps from this medium post and now it works:

https://medium.com/@jodywan/cloud-native-devops-08-harbor-2-0-with-traefik-2-0-reverse-proxy-5acbba95534a

**Comment by Dzhan85** [comment]
> @PauloFavero I followed the steps from this medium post and now it works:
> https://medium.com/@jodywan/cloud-native-devops-08-harbor-2-0-with-traefik-2-0-reverse-proxy-5acbba95534a
> 
> So relativeurls and external_url did the trick.
> 
> ```
> # http related config
> http:
>   # port for http, default is 80. If https enabled, this port will redirect to https port
>   port: 80
>   relativeurls: true
> ...
> # Uncomment external_url if you want to enable external proxy
> # And when it enabled the hostname will no longer used
> external_url: https://reg.mydomain.com
> ```

I tried in my case, it did not work for me.

**Comment by hostalp** [comment]
Did anyone manage to resolve/workaround it somehow?
I'm facing the same issue at Harbor 2.1.1 even when using HTTPS. The token gets generated, but then the request to registry fails with `error authenticating user "harbor_registry_user": authentication failure`
Only robot accounts work properly.

**Comment by burdzwastaken** [comment]
We are seeing the same behaviour trying to upgrade Harbor from v1.10.6 > v2.0.5. We are running Harbor behind an AWS ELB > nginx using the configmap-http.yaml configuration. We have external_url set. Authentication to the UI via OIDC works perfectly however via the docker CLI calls are failing with unauthorized

We have updated the registry configuration to be the following:
```
    auth:
      htpasswd:
        realm: harbor-registry-basic-realm
        path: /etc/registry/passwd
    compatibility:
      schema1:
        enabled: true
```

We have made sure the passwd username and password is shared with the right services (core/jobservice) and have it correctly mounted within the registry container. The error we see is the same as above (domain has been changed to share) from registry service:
```
time="2021-02-02T15:59:29.22386559Z" level=error msg="error authenticating user "harbor_registry_user": authentication failure" go.version=go1.14.7 http.request.host=my.registry.company.com http.request.id=044cf4e8-8720-4541-b9d5-38d21a4968f8 http.request.method=GET http.request.remoteaddr=54.245.130.173 http.request.uri="/v2/" http.request.useragent="docker/20.10.2 go/go1.13.15 git-commit/8891c58 kernel/4.19.0-13-amd64 os/linux arch/amd64 UpstreamClient(Docker-Client/20.10.2 \(linux\))" 
time="2021-02-02T15:59:29.223928923Z" level=warning msg="error authorizing context: basic authentication challenge for realm "harbor-registry-basic-realm": authentication failure" go.version=go1.14.7 http.request.host=my.registry.company.com http.request.id=044cf4e8-8720-4541-b9d5-38d21a4968f8 http.request.method=GET http.request.remoteaddr=54.245.130.173 http.request.uri="/v2/" http.request.useragent="docker/20.10.2 go/go1.13.15 git-commit/8891c58 kernel/4.19.0-13-amd64 os/linux arch/amd64 UpstreamClient(Docker-Client/20.10.2 \(linux\))"
```

Headers from our registry:
```
[] burdz@~: curl -X GET -I "https://my.registry.company.com/v2/"
HTTP/1.1 401 Unauthorized
Content-Type: application/json; charset=utf-8
Date: Wed, 10 Feb 2021 09:34:35 GMT
Docker-Distribution-Api-Version: registry/2.0
Server: nginx
Set-Cookie: sid=ea016b3b52ceb9dbb84cefb223b19d61; Path=/; HttpOnly
Www-Authenticate: Bearer realm="https://my.registry.company.com/service/token",service="harbor-registry"
X-Request-Id: 4744aab4-33f4-4ea2-a2c5-ed8ff5b54e66
Content-Length: 76
Connection: keep-alive
```

When we try to authenticate via the docker CLI:
```
[] burdz@~: docker login https://my.registry.company.com
Username: Matt_Burdan
Password: 
Error response from daemon: login attempt to https://harbor.kbuild-dev.msap.io/v2/ failed with status: 401 Unauthorized
```

**Comment by burdzwastaken** [comment]
@reasonerjt or anyone in the thread have you solved this? this is blocking our upgrade and we would like to get to 2.X

**Comment by llitfkitfk** [comment]
@tomhouweling1987 solved

**Comment by burdzwastaken** [comment]
@llitfkitfk I have that set and it has not changed the situation. I still see the same issue when trying to pull/push images.

**Comment by burdzwastaken** [comment]
commenting to remove the stale label. we are still facing this issue but will continue investigating more next week as so far none of the suggested solutions have resolved it

**Comment by PauloFavero** [comment]
We finally fixed it! We installed harbor 2.3 and we configured the harbor.yml as follows:

```
# Configuration file of Harbor

# The IP address or hostname to access admin UI and registry service.
# DO NOT use localhost or 127.0.0.1, because Harbor needs to be accessed by external clients.
hostname: our.domain.com

# http related config
http:
  # port for http, default is 80. If https enabled, this port will redirect to https port
  port: 80

# https related config
#https:
  # https port for harbor, default is 443
  #port: 443
  # The path of cert and key files for nginx
#  certificate: /your/certificate/path
#  private_key: /your/private/key/path

# # Uncomment following will enable tls communication between all harbor components
# internal_tls:
#   # set enabled to true means internal tls is enabled
#   enabled: true
#   # put your cert and key files on dir
#   dir: /etc/harbor/tls/internal

# Uncomment external_url if you want to enable external proxy
# And when it enabled the hostname will no longer used
external_url: our.domain.com

# The initial password of Harbor admin
# It only works in first time to install harbor
# Remember Change the admin password from UI after launching Harbor.
harbor_admin_password: admin

# Harbor DB configuration
database:
  # The password for the root user of Harbor DB. Change this before any production use.
  password: 123456
  # The maximum number of connections in the idle connection pool. If it <=0, no idle connections are retained.
  max_idle_conns: 100
  # The maximum number of open connections to the database. If it <= 0, then there is no limit on the number of open connections.
  # Note: the default number of connections is 1024 for postgres of harbor.
  max_open_conns: 900

# The default data volume
data_volume: /data

# Harbor Storage settings by default is using /data dir on local filesystem
# Uncomment storage_service setting If you want to using external storage
# storage_service:
#   # ca_bundle is the path to the custom root ca certificate, which will be injected into the truststore
#   # of registry's and chart repository's containers.  This is usually needed when the user hosts a internal storage with self signed certificate.
#   ca_bundle:

#   # storage backend, default is filesystem, options include filesystem, azure, gcs, s3, swift and oss
#   # for more info about this configuration please refer https://docs.docker.com/registry/configuration/
#   filesystem:
#     maxthreads: 100
#   # set disable to true when you want to disable registry redirect
#   redirect:
#     disabled: false

# Trivy configuration
#
# Trivy DB contains vulnerability information from NVD, Red Hat, and many other upstream vulnerability databases.
# It is downloaded by Trivy from the GitHub release page https://github.com/aquasecurity/trivy-db/releases and cached
# in the local file system. In addition, the database contains the update timestamp so Trivy can detect whether it
# should download a newer version from the Internet or use the cached one. Currently, the database is updated every
# 12 hours and published as a new release to GitHub.
trivy:
  # ignoreUnfixed The flag to display only fixed vulnerabilities
  ignore_unfixed: false
  # skipUpdate The flag to enable or disable Trivy DB downloads from GitHub
  #
  # You might want to enable this flag in test or CI/CD environments to avoid GitHub rate limiting issues.
  # If the flag is enabled you have to download the `trivy-offline.tar.gz` archive manually, extract `trivy.db` and
  # `metadata.json` files and mount them in the `/home/scanner/.cache/trivy/db` path.
  skip_update: false
  #
  # insecure The flag to skip verifying registry certificate
  insecure: false
  # github_token The GitHub access token to download Trivy DB
  #
  # Anonymous downloads from GitHub are subject to the limit of 60 requests per hour. Normally such rate limit is enough
  # for production operations. If, for any reason, it's not enough, you could increase the rate limit to 5000
  # requests per hour by specifying the GitHub access token. For more details on GitHub rate limiting please consult
  # https://developer.github.com/v3/#rate-limiting
  #
  # You can create a GitHub token by following the instructions in
  # https://help.github.com/en/github/authenticating-to-github/creating-a-personal-access-token-for-the-command-line
  #
  # github_token: xxx

jobservice:
  # Maximum number of job workers in job service
  max_job_workers: 10

notification:
  # Maximum retry count for webhook job
  webhook_job_max_retry: 10

chart:
  # Change the value of absolute_url to enabled can enable absolute url in chart
  absolute_url: disabled

# Log configurations
log:
  # options are debug, info, warning, error, fatal
  level: info
  # configs for logs in local storage
  local:
    # Log files are rotated log_rotate_count times before being removed. If count is 0, old versions are removed rather than rotated.
    rotate_count: 50
    # Log files are rotated only if they grow bigger than log_rotate_size bytes. If size is followed by k, the size is assumed to be in kilobytes.
    # If the M is used, the size is in megabytes, and if G is used, the size is in gigabytes. So size 100, size 100k, size 100M and size 100G
    # are all valid.
    rotate_size: 200M
    # The directory on your host that store log
    location: /var/log/harbor

  # Uncomment following lines to enable external syslog endpoint.
  # external_endpoint:
  #   # protocol used to transmit log to external endpoint, options is tcp or udp
  #   protocol: tcp
  #   # The host of external endpoint
  #   host: localhost
  #   # Port of external endpoint
  #   port: 5140

#This attribute is for migrator to detect the version of the .cfg file, DO NOT MODIFY!
_version: 2.3.0

# Uncomment external_database if using external database.
# external_database:
#   harbor:
#     host: harbor_db_host
#     port: harbor_db_port
#     db_name: harbor_db_name
#     username: harbor_db_username
#     password: harbor_db_password
#     ssl_mode: disable
#     max_idle_conns: 2
#     max_open_conns: 0
#   notary_signer:
#     host: notary_signer_db_host
#     port: notary_signer_db_port
#     db_name: notary_signer_db_name
#     username: notary_signer_db_username
#     password: notary_signer_db_password
#     ssl_mode: disable
#   notary_server:
#     host: notary_server_db_host
#     port: notary_server_db_port
#     db_name: notary_server_db_name
#     username: notary_server_db_username
#     password: notary_server_db_password
#     ssl_mode: disable

# Uncomment external_redis if using external Redis server
# external_redis:
#   # support redis, redis+sentinel
#   # host for redis: <host_redis>:<port_redis>
#   # host for redis+sentinel:
#   #  <host_sentinel1>:<port_sentinel1>,<host_sentinel2>:<port_sentinel2>,<host_sentinel3>:<port_sentinel3>
#   host: redis:6379
#   password:
#   # sentinel_master_set must be set to support redis+sentinel
#   #sentinel_master_set:
#   # db_index 0 is for core, it's unchangeable
#   registry_db_index: 1
#   jobservice_db_index: 2
#   chartmuseum_db_index: 3
#   trivy_db_index: 5
#   idle_timeout_seconds: 30

# Uncomment uaa for trusting the certificate of uaa instance that is hosted via self-signed cert.
# uaa:
#   ca_file: /path/to/ca

# Global proxy
# Config http proxy for components, e.g. http://my.proxy.com:3128
# Components doesn't need to connect to each others via http proxy.
# Remove component from `components` array if want disable proxy
# for it. If you want use proxy for replication, MUST enable proxy
# for core and jobservice, and set `http_proxy` and `https_proxy`.
# Add domain to the `no_proxy` field, when you want disable proxy
# for some special registry.
proxy:
  http_proxy:
  https_proxy:
  no_proxy:
  components:
    - core
    - jobservice
    - trivy

# metric:
#   enabled: false
#   port: 9090
#   path: /metrics
```

Thanks to @llitfkitfk ❤️
